cloud/blockstore/libs/nbd/netlink_device.cpp

```diff
@@ -0,0 +1,454 @@
+        , Disconnect(disconnect)
+    {
+        Log = Logging->CreateLog("BLOCKSTORE_NBD");
+
+        if (sscanf(DeviceName.c_str(), "/dev/nbd%u", &DeviceIndex) != 1) {
+            throw TServiceError(E_FAIL) << "invalid nbd device target";
+        }
+    }
+
+    ~TNetlinkDevice()
+    {
+        Stop();
+    }
+
+    void Start() override
+    {
+        ConnectSocket();
+        ConnectDeviceAsync();
+    }
+
+    void Stop() override
+    {
+        if (AtomicSwap(&ShouldStop, 1) == 1) {
+            return;
+        }
+
+        if (Disconnect) {
+            DisconnectDevice();
+            DisconnectSocket();
+        }
+    }
+
+private:
+    void ConnectSocket();
+    void DisconnectSocket();
+
+    void ConnectDevice(bool connected);
+    void ConnectDeviceAsync();
+    void DisconnectDevice();
+
+    static int StatusHandler(nl_msg* msg, void* arg);
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+void TNetlinkDevice::ConnectSocket()
+{
+    STORAGE_DEBUG("connect socket");
+
+    TSocket socket(ConnectAddress);
+    if (IsTcpAddress(ConnectAddress)) {
+        socket.SetNoDelay(true);
+    }
+
+    TSocketInput in(socket);
+    TSocketOutput out(socket);
+
+    Handler = CreateClientHandler(Logging);
+    Y_ENSURE(Handler->NegotiateClient(in, out));
+
+    Socket = socket;
+}
+
+void TNetlinkDevice::DisconnectSocket()
+{
+    STORAGE_DEBUG("disconnect socket");
+
+    Socket.Close();
+}
+
+void TNetlinkDevice::ConnectDevice(bool connected)
+{
+    auto cmd = NBD_CMD_CONNECT;
+
+    if (connected) {
+        if (!Reconfigure) {
+            throw TServiceError(E_FAIL) << DeviceName << " is busy";
+        }
+        STORAGE_INFO("reconfigure " << DeviceName);
+        cmd = NBD_CMD_RECONFIGURE;
+    } else {
+        STORAGE_INFO("connect " << DeviceName);
+    }
+
+    TNetlinkSocket nlsocket;
+
+    auto* msg = nlmsg_alloc();
+    if (msg == nullptr) {
+        throw TServiceError(E_FAIL) << "unable to allocate netlink message";
+    }
+    genlmsg_put(
+        msg,
+        NL_AUTO_PORT,
+        NL_AUTO_SEQ,
+        nlsocket.Family(),
+        0,
+        0,
+        cmd,
+        0);
+
+    const auto& info = Handler->GetExportInfo();
+    if (DeviceIndex >= 0) {
+        NLA_PUT_U32(msg, NBD_ATTR_INDEX, DeviceIndex);
+    }
+    NLA_PUT_U64(msg, NBD_ATTR_SIZE_BYTES, info.Size);
+    NLA_PUT_U64(msg, NBD_ATTR_BLOCK_SIZE_BYTES, info.MinBlockSize);
+    NLA_PUT_U64(msg, NBD_ATTR_SERVER_FLAGS, info.Flags);
+    if (Timeout) {
+        NLA_PUT_U64(msg, NBD_ATTR_TIMEOUT, Timeout.Seconds());
+    }
+    if (DeadConnectionTimeout) {
+        NLA_PUT_U64(
+            msg,
+            NBD_ATTR_DEAD_CONN_TIMEOUT,
+            DeadConnectionTimeout.Seconds());
+    }
+
+    auto* attr = nla_nest_start(msg, NBD_ATTR_SOCKETS);
+    if (!attr) {
+        throw TServiceError(E_FAIL) << "unable to nest NBD_ATTR_SOCKETS";
+    }
+    auto* opt = nla_nest_start(msg, NBD_SOCK_ITEM);
+    if (!opt) {
```

Comment:
missing `nlmsg_free()`

Reply:
Right... Might be a good idea to make a wrapper